WIP: Provide method addition

girdharshubham opened the pull request:
# Changelog
* provide method added for
  * Socket
  * SocketApp
  * HttpAttribute
  * Response
  * HttpApp
* Unit test case added for Socket's provide
* Unit test case added for HttpApp's provide
### Closes #477

girdharshubham committed:
feat(provide): provide method added for Socket for some cases

girdharshubham committed:
feat(provide): provide added for Socket, SocketApp, HttpAttribute, Response, and HttpApp

test(*): for provide. unit test case for socket seems flaky*

girdharshubham committed:
refactor(test): unused parameter warning removed

girdharshubham committed:
refactor(*): HttpAppSpec reformatted

tusharmath commented on zio-http/src/main/scala/zhttp/http/HttpAttribute.scala:
```diff
@@ -1,13 +1,20 @@
 private[zhttp] sealed trait HttpAttribute[-R, +E] extends Product with Serializable { self =>
   def exit: HttpAttribute.AttributeExit[R, E]                                   = HttpAttribute.exit(self)
   def ++[R1 <: R, E1 >: E](other: HttpAttribute[R1, E1]): HttpAttribute[R1, E1] = HttpAttribute.Combine(self, other)
+  def provide(r: R)(implicit env: NeedsEnv[R]): HttpAttribute[Any, E]           = self match {
```
Create a new type for `HttpAttribute.Provide`

tusharmath commented on zio-http/src/main/scala/zhttp/socket/Socket.scala:
```diff
@@ -20,6 +20,9 @@ sealed trait Socket[-R, +E, -A, +B] { self =>
   private[zhttp] def execute(a: A): ZStream[R, E, B] = self(a)
 
+  def provide(r: R)(implicit needsEnv: NeedsEnv[R]): Socket[Any, E, A, B] =
+    Socket.FromStreamingFunction(a => self(a).provide(r))
```
```suggestion
    Socket.Provide(r, self)
```

girdharshubham replied:
While providing an implementation for the Provide type and the method, the compiler keeps on warning that the effect has **Any** as its requirement
```scala
sealed trait Socket[-R, +E, -A, +B] { self =>
  def apply(a: A): ZStream[R, E, B] = self match {
    ...
    case Provide(m, r)               => m(a).provide(r)
  }
  def provide(r: R)(implicit needsEnv: NeedsEnv[R]): Socket[Any, E, A, B] = Provide(self, r)
}
object Socket {
  private final case class Provide[R, E, A, B](m: Socket[R, E, A, B], r: R)                 extends Socket[Any, E, A, B]
}
```

I believe the intention behind your comment is that with provide we should return a declaration of what we'd end up with and then the apply method is where the actual environment elimination should be.

girdharshubham replied:
This has been resolved in newer commit(s)

tusharmath commented on zio-http/src/main/scala/zhttp/socket/SocketApp.scala:
```diff
@@ -49,6 +50,23 @@ sealed trait SocketApp[-R, +E] { self =>
     loop(self, SocketConfig[R, E]())
   }
 
+  def provide(r: R)(implicit env: NeedsEnv[R]): SocketApp[Any, E] =
```
```suggestion
  def provide(r: R)(implicit env: NeedsEnv[R]): SocketApp[Any, E] = Provide(r, self)
```

girdharshubham committed:
refactor(provide): Socket.scala's provide implemented as per PR comment
test(provide): Unit test case for provide written in SocketSpec.scala
wip(SocketApp)

girdharshubham committed:
SimpleSocketProvideServer added

girdharshubham commented:
Closing this PR in favour of smaller scoped PRs.
Comments will be resolved in them.
